### PR TITLE
4099 ES6 Refactor lib/HotModuleReplacementPlugin

### PR DESCRIPTION
**Pull request by KTruong888**
**What kind of change does this PR introduce?**

ES6 Refactor of HotModuleReplacementPlugin

**Did you add tests for your changes?**

Tests were already written.

**Summary**

Issue:
https://github.com/webpack/webpack/issues/4099

**Other Information**

Still working on this PR but I'm a bit stuck and looking for any advice. I managed to fix a lot of the broken tests as a result of the refactor, except for the last few that all seem to stem from the same problem. 

All of the broken tests say something along the lines of "TestCases all-combined chunks .... should compile", along with red text underneath:

Unexpected token: keyword (const) [bundle.js:218,16]

I tried changing everything from const to let just to see how it would go, but then it had issues in the same test file with 'chunkId'.

Does anyone know how to run this specific test (TestCases.test.js all-combined) instead of having to run the whole test suite, so I can better debug it? 

I tried `$(npm bin)/mocha --grep TestCases`, but it takes longer to run that single command than it does running the whole test suite because it keeps pausing and giving me weird messages like this:

HEAD 2881ms ± 306ms [2742ms; 3020ms]
․        v3.0.0-rc.2 2973ms ± 338ms [2820ms; 3127ms]
․․        HEAD 1462ms ± 90ms [1431ms; 1493ms]
․

### CODE REVIEW

File: lib/HotModuleReplacementPlugin.js [hunk 1]
```diff
@@ -3,260 +3,249 @@
...
+			compilation.mainTemplate.plugin("bootstrap", function(source, chunk, hash) {
+				source = this.applyPluginsWaterfall("hot-bootstrap", source, chunk, hash);
+				return this.asString([
+					source,
+					"",
+					hotInitCode
+					.replace(/\$require\$/g, this.requireFn)
+					.replace(/\$hash\$/g, JSON.stringify(hash))
+					.replace(/\$requestTimeout\$/g, requestTimeout)
+					.replace(/\/\*foreachInstalledChunks\*\//g, chunk.chunks.length > 0 ? "for(const chunkId in installedChunks)" : `const chunkId = ${JSON.stringify(chunk.id)};`)
```
Comment (sokra) [inline lib/HotModuleReplacementPlugin.js]: `const chunkId` -> `var chunkId`

This is code which is in the generated bundle. The generated code should not contain higher ES.

File: lib/HotModuleReplacementPlugin.js [hunk 2]
```diff
@@ -3,260 +3,249 @@
...
-				.replace(/\$hash\$/g, JSON.stringify(hash))
-				.replace(/\$requestTimeout\$/g, requestTimeout)
-				.replace(/\/\*foreachInstalledChunks\*\//g, chunk.chunks.length > 0 ? "for(var chunkId in installedChunks)" : "var chunkId = " + JSON.stringify(chunk.id) + ";")
-			]);
-		});
+			compilation.mainTemplate.plugin("require-extensions", function(source) {
+				const buf = [source];
+				buf.push("");
+				buf.push("// __webpack_hash__");
+				buf.push(`${this.requireFn}.h = () => hotCurrentHash`);
```
Comment (sokra) [inline lib/HotModuleReplacementPlugin.js]: don't refactor the generated code.

Comment (KTruong888) [inline lib/HotModuleReplacementPlugin.js]: stopped refactoring the generated code on line 156

**Comment by sokra** [comment]
```
  2) TestCases all-combined chunks context should compile:
     Error: Errors while compiling:

bundle.js from UglifyJs
Unexpected token: punc ()) [bundle.js:726,35]
```

It looks like there is more generated code in ES6 format.

This is how to find it:

* run the test locally: `mocha -g "all-combined chunks context"`
* open the file `test/js/all-combined/chunks/context/bundle.js`
* check the line from the error `bundle.js:726,35` -> line 726, column 35
* find the origin of the generated code in the HotModuleReplacementPlugin (or paste it here)
* change back to ES3 code.

**Comment by sokra** [comment]
> 2) TestCases all-combined chunks context should compile:
>      Error: Errors while compiling:
> 
> bundle.js from UglifyJs
> Unexpected token: punc ()) [bundle.js:726,35]
> It looks like there is more generated code in ES6 format.
> 
> This is how to find it:
> 
> run the test locally: mocha -g "all-combined chunks context"
> open the file test/js/all-combined/chunks/context/bundle.js
> check the line from the error bundle.js:726,35 -> line 726, column 35
> find the origin of the generated code in the HotModuleReplacementPlugin (or paste it here)
> change back to ES3 code.

Sorry, I was wrong. It's all fine now.

**Comment by sokra** [comment]
Thanks

**Pull request merged**